### PR TITLE
Strengthen assertions in trace validation

### PR DESCRIPTION
**Pull request by lemmy**
Related https://github.com/microsoft/CCF/pull/6000

### CODE REVIEW

**Comment by lemmy** [comment]
This PR "just" adds assertions that don't make any of the existing scenarios fail.  We should get this merged to avoid future regressions, while we keep digging why the assertions don't generally hold.

**Review by heidihoward — APPROVED** [review]

**Comment by heidihoward** [comment]
Aside: I was surprised to see a 11/14% increase in the number of states for model checking with reconfiguration. This PR didn't seem to touch the model checking at all

File: tla/consensus/Traceccfraft.tla [hunk 1]
```diff
@@ -147,12 +155,20 @@ IsTimeout ==
     /\ logline.msg.state.leadership_state = "Candidate"
     /\ Timeout(logline.msg.state.node_id)
     /\ Range(logline.msg.state.committable_indices) \subseteq CommittableIndices(logline.msg.state.node_id)
+    /\ commitIndex[logline.msg.state.node_id] = logline.msg.state.commit_idx
```
Comment (achamayou) [inline tla/consensus/Traceccfraft.tla:158]: This is very repetitive, I would have preferred if it had been factored out.

Comment (lemmy) [reply tla/consensus/Traceccfraft.tla:158]: Absolutely, I even started with an `IsState` definition that asserted `log_index`, `membership_state`, `leadership_state`, and` `commit_index`.  I decided to inline the four assertions for as long as not all actions are consistent.

**Comment by lemmy** [comment]
> Aside: I was surprised to see a 11/14% increase in the number of states for model checking with reconfiguration. This PR didn't seem to touch the model checking at all

Where do I find the numbers that come out of TLC and go into cimetrics?

**Comment by achamayou** [comment]
@lemmy in the TLC section of the (admittedly large) ccf-bot post on every PR:

image

**Comment by lemmy** [comment]
@achamayou I meant the numbers that go into the chart generator. As @heidihoward observed, this commit should not cause a change in state-space statistics, and I'd like to check if there is a bug in TLC or cimetrics.

**Comment by eddyashton** [comment]
I think this increase was not from this PR, but from something earlier. cimetrics plots a moving average, so in this case we see a plot halfway through `main`'s average being updated (ie, after only a few runs on `main`, where the average hasn't settled to the new value yet).

**Comment by achamayou** [comment]
@lemmy sim https://github.com/microsoft/CCF/blob/f7dab4a9b7fbb665ce9206d6ffcd1babe7d00d7c/.azure-pipelines-templates/simulation.yml#L31
mc https://github.com/microsoft/CCF/blob/f7dab4a9b7fbb665ce9206d6ffcd1babe7d00d7c/.azure-pipelines-templates/model_checking.yml#L36

**Comment by achamayou** [comment]
As @eddyashton says, the line is an exponentially weighted moving average, and so takes some time to catch up. The small blue dots are the actual data.

**Comment by eddyashton** [comment]
#5978 is the PR that actually increased these numbers.

**Pull request merged**